### PR TITLE
br: fix renaming pvc when there are multiple volumes in tikv pod

### LINKED ISSUE #4950: TiKV crash when execute EBS restore, and every tikv has multiple volumes
## Bug Report

**What version of Kubernetes are you using?**

Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.5", GitCommit:"5c99e2ac2ff9a3c549d9ca665e7bc05a3e18f07e", GitTreeState:"clean", BuildDate:"2021-12-16T08:38:33Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.17-eks-48e63af", GitCommit:"47b89ea2caa1f7958bc6539d6865820c86b4bf60", GitTreeState:"clean", BuildDate:"2023-01-24T09:34:06Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}

**What version of TiDB Operator are you using?**

v1.4.4

**What storage classes exist in the Kubernetes cluster and what are used for PD/TiKV pods?**

gp2

**What's the status of the TiDB cluster pods?**

**What did you do?**

1. deploy a TiDB cluster, and every tikv mounts 3 volumes.
2. do an EBS backup
3. do an EBS restore

**What did you expect to see?**

EBS restore successfully

**What did you see instead?**

tikv pods always crash

### PR DESCRIPTION
**Pull request by WangLe1321**
### What problem does this PR solve?

Fix the bug that TiKV pods crash when execute EBS restore to restore TiKV with multiple volumes for the reason that we renamed pvc of TiKV by mistake.

Closes #4950 

### What is changed and how does it work?

The root cause is that we suppose there is only one volume in one TiKV when we sequence the pvc. When there are multiple volumes in one TiKV, the sequence of pvc will be disordered, which results in TiKV pods and restored pvc can't match.

So we group pvc of TiKV by volume name, we sequence pvc group by group.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: pkg/backup/snapshotter/snapshotter.go [hunk 1]
```diff
@@ -508,64 +508,76 @@ func (m *StoresMixture) ProcessCSBPVCsAndPVs(r *v1alpha1.Restore, csb *CloudSnap
...
 		// subMatches contains full text that matches regex and the matches in brackets.
-		// so if the pvc matches regex, it should contain 2 items.
-		if len(subMatches) != 2 {
+		// so if the pvc matches regex, it should contain 3 items.
+		if len(subMatches) != 3 {
 			return nil, nil, fmt.Errorf("pvc name %s doesn't match regex %s", pvc.Name, reStr)
 		}
 		// get the number of pvc, for example, there is a pvc "tikv-db-tikv-0", try to get the number "0"
-		index, err := strconv.Atoi(subMatches[1])
+		index, err := strconv.Atoi(subMatches[2])
```
Comment (grovecai) [inline pkg/backup/snapshotter/snapshotter.go:533]: pls name subMatches[2], make it easy to read

File: pkg/backup/snapshotter/snapshotter.go [hunk 2]
```diff
@@ -508,64 +508,76 @@ func (m *StoresMixture) ProcessCSBPVCsAndPVs(r *v1alpha1.Restore, csb *CloudSnap
...
+// [tikv-db-tikv-0, tikv-db-tikv-1, tikv-db-tikv-2, tikv-raft-db-tikv-0, tikv-raft-db-tikv-1, tikv-raft-db-tikv-2]
 func resetPVCSequence(stsName string, pvcs []*corev1.PersistentVolumeClaim, pvs []*corev1.PersistentVolume) (
 	[]*corev1.PersistentVolumeClaim, []*corev1.PersistentVolume, error) {
 	type indexedPVC struct {
 		index int
 		pvc   *corev1.PersistentVolumeClaim
 	}
-	indexedPVCs := make([]*indexedPVC, 0, len(pvcs))
-	reStr := fmt.Sprintf(`%s-(\d+)$`, stsName)
+	indexedPVCsGroups := make(map[string][]*indexedPVC, 8)
```
Comment (Ehco1996) [inline pkg/backup/snapshotter/snapshotter.go:522]: why hardcode to 8?

Comment (WangLe1321) [reply pkg/backup/snapshotter/snapshotter.go:522]: The length of this map depends on the count of volumes attached in TiKV, and we don't know how many volumes in a TiKV before we group them. I think 8 is a big enough number to make the map not expansion.

File: pkg/backup/snapshotter/snapshotter.go [hunk 3]
```diff
@@ -508,64 +508,76 @@ func (m *StoresMixture) ProcessCSBPVCsAndPVs(r *v1alpha1.Restore, csb *CloudSnap
...
+	reStr := fmt.Sprintf(`^(.+)-%s-(\d+)$`, stsName)
 	re := regexp.MustCompile(reStr)
 	for _, pvc := range pvcs {
 		subMatches := re.FindStringSubmatch(pvc.Name)
 		// subMatches contains full text that matches regex and the matches in brackets.
-		// so if the pvc matches regex, it should contain 2 items.
-		if len(subMatches) != 2 {
+		// so if the pvc matches regex, it should contain 3 items.
+		if len(subMatches) != 3 {
```
Comment (StepOnce7) [inline pkg/backup/snapshotter/snapshotter.go:529]: not 4? "tikv-db-tikv-0"

**Review by StepOnce7 — APPROVED** [review]

File: pkg/backup/snapshotter/snapshotter.go [hunk 4]
```diff
@@ -508,64 +508,76 @@ func (m *StoresMixture) ProcessCSBPVCsAndPVs(r *v1alpha1.Restore, csb *CloudSnap
...
 		// get the number of pvc, for example, there is a pvc "tikv-db-tikv-0", try to get the number "0"
-		index, err := strconv.Atoi(subMatches[1])
+		index, err := strconv.Atoi(subMatches[2])
 		if err != nil {
 			return nil, nil, fmt.Errorf("parse index %s of pvc %s to int: %s", subMatches[1], pvc.Name, err.Error())
 		}
-		indexedPVCs = append(indexedPVCs, &indexedPVC{
+		// get the volume name of pod from pvc, for example, there is a pvc "tikv-db-tikv-0", get "tikv"
+		// there can be multiple volumes in a pod, so we should group pvc by volume name
+		indexedPVCs, ok := indexedPVCsGroups[subMatches[1]]
```
[the remark this answers is not in the document]
Comment (WangLe1321) [reply pkg/backup/snapshotter/snapshotter.go]: done

**Review by grovecai — APPROVED** [review]

**Review by Ehco1996 — APPROVED** [review]

**Review by csuzhangxc — APPROVED** [review]

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**